### PR TITLE
KMeans Sparse Init Update

### PR DESCRIPTION
**Pull request by md-shafiul-alam**
# Description
The pull request fixes the difference in centroids between dense and sparse method for KMeans++ init method. 

Changes proposed in this pull request:
- bug fix daal kmeans++ init for sparse data 
- allow oneDAL to pass nTrials to match with Scikit-learn

### CODE REVIEW

File: cpp/daal/src/algorithms/kmeans/kmeans_plusplus_init_impl.i [hunk 1]
```diff
@@ -292,15 +292,22 @@ public:
...
+
         for (size_t iRow = 0u; iRow < nRowsToProcess; iRow++)
         {
-            algorithmFPType dist2 = algorithmFPType(0);
+            algorithmFPType dist2 = pLastAddedCenterSumSq;
             const size_t nValues  = rowIdx[iRow + 1] - rowIdx[iRow];
             for (size_t i = 0u; i < nValues; i++, csrCursor++)
             {
-                dist2 += (pData[csrCursor] - pLastAddedCenter[colIdx[csrCursor] - 1]) * (pData[csrCursor] - pLastAddedCenter[colIdx[csrCursor] - 1]);
+                dist2 += pData[csrCursor] * pData[csrCursor];
```
Comment (Alexandr-Solovev) [inline cpp/daal/src/algorithms/kmeans/kmeans_plusplus_init_impl.i:309]: probably, its better to merge these operations in one line

File: cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp [hunk 2]
```diff
@@ -43,9 +45,15 @@ static compute_result<Task> call_daal_kernel(const context_cpu& ctx,
     const std::int64_t column_count = data.get_column_count();
     const std::int64_t cluster_count = desc.get_cluster_count();
 
+    const std::int64_t init_trial_count = desc.get_local_trials_count();
+    const auto additional = std::log(cluster_count);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp:49]: It can be probably replaced with sycl::log to avoid #include <cmath>

Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp:49]: just a suggestion: probably its better to replace with

```
std::int64_t trial_count = desc.get_local_trials_count();
if(trial_count== -1)
{
const auto additional = std::log(cluster_count);
trial_count = 2 + std::int64_t(additional);
}
```
to avoid additional computations in case when trial count is set

File: cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp [hunk 3]
```diff
@@ -43,9 +45,17 @@ static compute_result<Task> call_daal_kernel(const context_cpu& ctx,
     const std::int64_t column_count = data.get_column_count();
     const std::int64_t cluster_count = desc.get_cluster_count();
 
+    std::int64_t trial_count = desc.get_local_trials_count();
+    if (trial_count == -1)
+    {
+        const auto additional = std::log(cluster_count);
+        trial_count = 2 + std::int64_t(additional);
+    }
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp:53]: Please add the comment describing the logic behind this. Maybe the link to Sklearn docs.

File: cpp/daal/src/algorithms/kmeans/kmeans_plusplus_init_impl.i [hunk 4]
```diff
@@ -292,15 +292,21 @@ public:
                                             const algorithmFPType * const pLastAddedCenter, const algorithmFPType * const aWeights,
                                             const algorithmFPType * const pDistSqBest)
     {
-        algorithmFPType sumOfDist2 = algorithmFPType(0);
-        size_t csrCursor           = 0u;
+        algorithmFPType sumOfDist2            = algorithmFPType(0);
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/kmeans/kmeans_plusplus_init_impl.i:295]: Please add the description to updateMinDistForITrials function.
I understand that it was not there before, but I hope that by adding couple of comments at a time we can make oneDAL's code more readable.

File: cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp [hunk 5]
```diff
@@ -43,9 +45,18 @@ static compute_result<Task> call_daal_kernel(const context_cpu& ctx,
...
     const std::int64_t cluster_count = desc.get_cluster_count();
 
+    //number of trials to pick each centroid from, 2 + int(ln(cluster_count)) works better than vanilla kmeans++
+    //https://github.com/scikit-learn/scikit-learn/blob/a63b021310ba13ea39ad3555f550d8aeec3002c5/sklearn/cluster/_kmeans.py#L108
+    std::int64_t trial_count = desc.get_local_trials_count();
+    if (trial_count == -1) {
+        const auto additional = std::log(cluster_count);
+        trial_count = 2 + std::int64_t(additional);
+    }
+
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp:55]: What kind of improvements this give? It seems this changes original behavior. Please also reflect it in the documentation if not already done.

**Review by samir-nasibli — COMMENTED** [review]
@md-shafiul-alam it is labeled like bug. Is it buggy or just fixes the bug?

Unfortunately we don't have good description for some labels, but should be addressed further.

File: cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp (hunk 5 above)
Comment (md-shafiul-alam) [reply cpp/oneapi/dal/algo/kmeans_init/backend/cpu/compute_kernel_dense.cpp:55]: This is actually an original behavior in daal and oneDAL distributed, somehow was missed on oneDAL CPU

**Comment by md-shafiul-alam** [comment]
> @md-shafiul-alam it is labeled like bug. Is it buggy or just fixes the bug?
> 
> Unfortunately we don't have good description for some labels, but should be addressed further.

This fixes the bug

**Comment by md-shafiul-alam** [comment]
Changes have been incorporated in [PR#2815](https://github.com/oneapi-src/oneDAL/pull/2815)
